src/main/kotlin/org/jitsi/nlj/transform/node/AudioRedHandler.kt

```diff
@@ -0,0 +1,267 @@
+                    RedPolicy.PROTECT_ALL -> true
+                    // RedPolicy.PROTECT_DOMINANT -> =isDominant
+                }
+            }
+
+            val audioRtpPacket = packetInfo.packetAs<AudioRtpPacket>()
+
+            sentAudioCache.insert(packetInfo.packetAs())
+
+            if (encapsulate) {
+                val redundancy = mutableListOf<RtpPacket>()
+                if (distance == RedDistance.TWO) {
+                    if (redundancy.maybeAddPacket(applySequenceNumberDelta(audioRtpPacket.sequenceNumber, -2))) {
+                        stats.redundancyPacketAdded()
+                    }
+                }
+                if (redundancy.maybeAddPacket(applySequenceNumberDelta(audioRtpPacket.sequenceNumber, -1))) {
+                    stats.redundancyPacketAdded()
+                }
+
+                val redPacket = RedAudioRtpPacket.builder.build(redPayloadType!!, audioRtpPacket, redundancy)
+                packetInfo.packet = redPacket
+
+                redundancy.forEach { BufferPool.returnBuffer(it.buffer) }
+                BufferPool.returnBuffer(audioRtpPacket.buffer)
+
+                stats.audioPacketEncapsulated()
+            } else {
+                stats.audioPacketForwarded()
+            }
+            return packetInfo
+        }
+
+        private fun MutableList<RtpPacket>.maybeAddPacket(seq: Int): Boolean {
+            if (currentLoss() < lossThreshold) {
+                return false
+            }
+
+            sentAudioCache.get(seq)?.item?.let {
+                // In vad-only mode, we only add redundancy for packets that have an audio level extension with the VAD
+                // bit set.
+                if (!vadOnly || getVad(it)) {
+                    add(it)
+                    return true
+                }
+            }
+            return false
+        }
+
+        private fun getVad(packet: RtpPacket): Boolean = audioLevelExtId?.let { extId ->
+                packet.getHeaderExtension(extId)?.let { AudioLevelHeaderExtension.getVad(it) } ?: false
+            } ?: false
+
+        /**
+         * Process an incoming RED packet. Depending on the configured policy and whether the receiver supports the
+         * RED format, it is either forwarded as it is or it is "stripped" to its primary encoding, with redundancy
+         * blocks being read if there are non-received packets.
+         */
+        fun transformRed(packetInfo: PacketInfo): List<PacketInfo> {
+            // Whether we need to strip the RED encapsulation
+            val strip = when (redPayloadType) {
+                null -> true
+                else -> when (policy) {
+                    RedPolicy.STRIP -> true
+                    // RedPolicy.PROTECT_DOMINANT -> !isDominant
+                    RedPolicy.NOOP, RedPolicy.PROTECT_ALL -> false
+                }
+            }
+
+            return if (strip) mutableListOf<PacketInfo>().apply {
```

Comment:
There's `buildList` for this use case, I think.

Reply:
It's an experimental API. Tests pass. Do you think it's safe to use?

Reply:
I think it's 100% safe--the experimental just refers to potential ABI breakages at some point, but the warnings can be a bit annoying.  I didn't remember it was experimental...we can leave it as-is.